### DISCUSSION TITLE
Universal Gcode Sender error "Grbl has not finished booting"

### DISCUSSION THREAD

**Discussion by jrdarrah** [terjeio/grblHAL#199] [Q&A]
I finally worked out the IP routing problem and can talk to the grblHAL board from UGS running on a Raspberry PI 4.  I do get the same error when connecting via the USB serial connection

[MSG:TELNET STREAM ACTIVE]
**** Connected to 10.1.1.2 @ 23 baud ****

When I enter any  commands I get the "Grbl has not finished booting" error. 

The UGS FAW section  has this

"Grbl has not finished booting."
This happens when UGS connects to a serial port and does not receive the GRBL startup string. Typically this is caused by a configuration problem and can be solved by one of the following:

Check the baud rate is 115200, or 9600 for very old versions of grbl.
Make sure you are connecting to the correct port.
Make sure you have installed any drivers required for your controller.
Make sure GRBL is properly flashed on your controller.

I'm compiling grblHAL so I can get the correct IP address so changing something else won't be a problem.  How do I get grblHAL to return the string  that UGS is waiting on?

**Comment by terjeio** [comment]
I'm compiling grblHAL so I can get the correct IP address so changing something else won't be a problem. How do I get grblHAL to return the string that UGS is waiting on?

By setting [compatibility level](https://github.com/terjeio/grblHAL/wiki/Compatibility-level) to a value that the UGS version you have is happy with. Start with 10, may work with 2.

**Reply by jrdarrah** [reply]
I tried them all 1 -> 10 and still get the same error. In the  UGS message log it says that it couldn't find a firmware with $22 key whatever that is.  I was able to get the code to send [Grbl 1.1f] still got the Grbl has not finished booting error.  In looking in GitHub I saw a development request for UGS to support grblHAL but no reply.  Tomorrow I'm going to try again with the USB serial connection. Has anyone gotten UGS talking to grblHAL?

**Comment by phil-barrett** [comment]
I've used a recent UGS with compat level 0 via USB on a RasPi.  Can you connect to grblHAL with a terminal app?  I used /dev/ACM0   Have not tried telnet.

Also, I've found that sometimes the RasPi or UGS doesn't like the USB connection - try using a different USB port.  I just tried it a minute ago and it wouldn't connect.  Moved the USB connection to a different port and it now works. I would get that working first before moving on to ethernet.

**Comment by breiler** [comment]
Hi @jrdarrah,

UGS will work with compat level 0 using USB if you use the latest version v.2.0.7. It isn't perfect but it should be able to connect and use most features. I have not tested the TCP connection yet.

The error you got "Grbl has not finished booting", is because it expects the welcome string "Grbl" or "GrblHAL xxxx" when connecting to the hardware. To make sure this welcome message is sent, UGS will attempt to reset the controller after connecting by sending "0x18". This maybe resets the network stack as well on GrblHAL?

We have discussed a new way of identifying which hardware variant UGS connects to, but this requires a bit of work which I have not yet prioritised.

**Reply by terjeio** [reply]
>  ...by sending "0x18". This maybe resets the network stack as well on GrblHAL?

It does not, only a hard reset does.

How to specify a network connection in the latest version? I have tried some variants without success.

**Reply by breiler** [reply]
The workflow for using TCP connections is still really non intuitive in UGS, I think that it was a hack that was shoehorned in a couple of years ago... =(

Very few people have used it until you guys (GrblHAL and Grbl_Esp32) showed up and I got to try that feature for myself the first time now in the Christmas holidays.

Select the TCP driver. In the connection field type the IP-address in the field for Port. And then the TCP port number in the Baud.
![2021-01-11 11-33-46 2021-01-11 11_36_23](https://user-images.githubusercontent.com/8962024/104170344-44266700-5401-11eb-9a38-1170ac5e2c7e.gif)

**Reply by terjeio** [reply]
Thanks. I can now connect.

A soft reset is sent and the controller replies with the boot message. Then silence fom UGS, waiting for more?
If I send a real-time request from a serial terminal then UGS wakes up and seems to work correctly.

If I push the boot message on connect UGS works after complaining about an unexpected ok message. This is due to a $$ request beeing interrupted by the soft reset:
```
GrblHAL 1.1f ['$' or '$HELP' for help]
>>> $$
GrblHAL 1.1f ['$' or '$HELP' for help]
>>> $G
>>> $$
$1 = 25    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
...
$307 = 80   
ok
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
ok
>>> $G
$0 = 10.0    (Step pulse time, microseconds)
$1 = 25    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
...
$305 = 23   
$307 = 80   
ok
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[Error] Error while processing response <ok>: An unexpected command was completed by the controller.
```

**Reply by jrdarrah** [reply]
Before I give up on the IP connection for now I recompiled with compatibility level = 0 and used telnet  to communicate with the grblHAL code. Based on the 0 compat level I would have expected that grblHAL would be sent. My plan as I wait on the mill to arrive is to get the software working both on the Raspberry and grblHAL on the breakout board. To that end, the breakout board has nothing atttached. Would that cause this error? If so I'll stand down until I can add the hardware components. Why didn't the grblHAL string come back first thing?

pi@CNCServer:~ $ telnet 10.1.1.2
Trying 10.1.1.2...
Connected to 10.1.1.2.
Escape character is '^]'.
[MSG:TELNET STREAM ACTIVE]
ALARM:10
[MSG:Emergency stop]
<Alarm|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,0|Pn:XYZHSE|Ov:100,100,100>

**Reply by terjeio** [reply]
> To that end, the breakout board has nothing atttached. Would that cause this error?

You mean alarm:10? Check [this page](https://github.com/terjeio/grblHAL/wiki/First-Run-Grbl-Settings).

> Why didn't the grblHAL string come back first thing?

You may argue that the greeting message should be sent on connect but I have decided against it since there are grblHAL systems where a connect event is not possible to detect. A sender should thus either send a soft reset to get it or try querying the controller state before deciding what to do. Personally I do not like the soft reset approach as this may have side-effects. See issue #108 for a discussion related to this.

**Comment by jrdarrah** [comment]
Thanks for all the help. UGS-grblHAL works with compatibility level 0 but only on USB-Serial. I'll keep looking to see when the ethernet connection is supported.

**Reply by breiler** [reply]
I added a step for running it over TCP here:
https://github.com/winder/Universal-G-Code-Sender/issues/1468

**Reply by jrdarrah** [reply]
I’m missing something. I don’t see anything about a new step at the link you posted in this thread

**Reply by breiler** [reply]
It is in the checklist of the issue: `Connection over TCP`
![image](https://user-images.githubusercontent.com/8962024/104226642-6c858400-5448-11eb-88a4-9b455138289e.png)

**Reply by terjeio** [reply]
I have now done some testing of network streaming, longest job was over 3 hours. No problems at all except for having to "kick" the connection alive via a real-time request over a serial connection.
